From 398c4c248b343661e8b48968d090cdfb5a1a601c Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@marvell.com>
Date: Sat, 5 Jan 2019 16:04:22 +0530
Subject: [PATCH 004/767] pci: octeontx2: Disable inbound write merging

commit 6db622b7627312261335dde698323b83bf07bd26 from
git@git.assembla.com:cavium/WindRiver.linux.git

Due to a PEM hardware issue, inbound write merging may cause
undefined operation. Hence disabling it.

Change-Id: I5626f777a2c39bdaf438fac454762c8e329f3c87
Signed-off-by: Sunil Goutham <sgoutham@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/7381
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/pci/controller/pci-octeontx2-pem.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/drivers/pci/controller/pci-octeontx2-pem.c b/drivers/pci/controller/pci-octeontx2-pem.c
index ddd2c62ec1ab..90b5537e8cdb 100644
--- a/drivers/pci/controller/pci-octeontx2-pem.c
+++ b/drivers/pci/controller/pci-octeontx2-pem.c
@@ -23,8 +23,9 @@
 /* Bridge config space reads/writes done using
  * these registers.
  */
-#define PEM_CFG_WR			0x18
-#define PEM_CFG_RD			0x20
+#define PEM_CFG_WR			0x018
+#define PEM_CFG_RD			0x020
+#define PEM_IB_MERGE_TIMER_CTL		0x1C0
 
 #define PCIERC_RAS_EINJ_EN		0x348
 #define PCIERC_RAS_EINJ_CTL6CMPP0	0x364
@@ -396,6 +397,7 @@ static int octeontx2_pem_init(struct device *dev, struct pci_config_window *cfg,
 {
 	struct octeontx2_pem_pci *pem_pci;
 	resource_size_t bar4_start;
+	u64 val;
 
 	pem_pci = devm_kzalloc(dev, sizeof(*pem_pci), GFP_KERNEL);
 	if (!pem_pci)
@@ -405,6 +407,15 @@ static int octeontx2_pem_init(struct device *dev, struct pci_config_window *cfg,
 	if (!pem_pci->pem_reg_base)
 		return -ENOMEM;
 
+	/* As per HW Errata 34726, an issue exists whereby inbound write
+	 * merging may cause undefined operation. Hence disabling it.
+	 *
+	 * Need to revisit this for future silicon passes and versions.
+	 */
+	val = readq(pem_pci->pem_reg_base + PEM_IB_MERGE_TIMER_CTL);
+	val |= BIT_ULL(10);
+	writeq(val, pem_pci->pem_reg_base + PEM_IB_MERGE_TIMER_CTL);
+
 	/*
 	 * The MSI-X BAR for the PEM and AER interrupts is located at
 	 * a fixed offset from the PEM register base.  Generate a
-- 
2.31.1

